runtime/shared/src/main/scala/org/http4s/googleapis/runtime/auth/IdentityPoolCredentials.scala

```diff
@@ -0,0 +1,270 @@
+  }
+
+  /** Create Google credentials from remote source. The main use-case of this credentials is
+    * Workload Identity Federation.
+    */
+  private[auth] def fromURL[F[_]](
+      client: Client[F],
+      projectId: String,
+      impersonationURL: Option[Uri],
+      urlSource: ExternalCredentialSource.Url,
+      externalAccount: CredentialsFile.ExternalAccount,
+      scopes: Seq[String],
+  )(implicit F: Temporal[F]): F[GoogleCredentials[F]] = (urlSource, impersonationURL) match {
+    case (u, Some(url)) => withImpersonation(client, projectId, url, u, externalAccount, scopes)
+    case (u, None) => withoutImpersonation(client, projectId, u, externalAccount, scopes)
+  }
+
+  private[auth] def forAWS[F[_]](
+      client: Client[F],
+      projectId: String,
+      impersonationURL: Option[Uri],
+      urlSource: ExternalCredentialSource.Aws,
+      externalAccount: CredentialsFile.ExternalAccount,
+      scopes: Seq[String],
+  )(implicit F: Temporal[F]): F[GoogleCredentials[F]] = ???
+
+  private def withoutImpersonation[F[_]: Files](
+      client: Client[F],
+      pid: String,
+      fileSource: ExternalCredentialSource.File,
+      externalAccount: CredentialsFile.ExternalAccount,
+      scopes: Seq[String],
+  )(implicit F: Temporal[F]): F[GoogleCredentials[F]] = fileSource match {
+    case ExternalCredentialSource.File(file, format) =>
+      F.ifM(Files[F].exists(Path(file)))(
+        F.unit,
+        F.raiseError(new IOException(s"$file not found")),
+      ).flatMap(_ =>
+        withoutImpersonation(
+          client,
+          pid,
+          subjectTokenFromFile(file, format),
+          externalAccount,
+          scopes,
+        ),
+      )
+  }
+  private def withoutImpersonation[F[_]](
+      client: Client[F],
+      pid: String,
+      urlSource: ExternalCredentialSource.Url,
+      externalAccount: CredentialsFile.ExternalAccount,
+      scopes: Seq[String],
+  )(implicit F: Temporal[F]): F[GoogleCredentials[F]] = urlSource match {
+    case ExternalCredentialSource.Url(url, headers, format) =>
+      F.fromEither(Uri.fromString(url))
+        .flatMap(url =>
+          withoutImpersonation(
+            client,
+            pid,
+            subjectTokenFromUrl(client, url, headers, format),
+            externalAccount,
+            scopes,
+          ),
+        )
+  }
+
+  /** The shared routine to fetch external credentials without service account impersonation.
+    *
+    * @param id
+    *   GCP project id
+    * @param retrieveSubjectToken
+    *   abstract platform specific actions to retrieve subject token for token exchange
+    * @param externalAccount
+    *   configuration for external account credentials generation
+    * @param scopes
+    *   scopes for authorization
+    *
+    * @return
+    *   a sts token for authorization skipping impersonation flow
+    */
+  private def withoutImpersonation[F[_]](
```

Comment:
This is the shared flow to get GoogleCredentials __without__ service account impersonation.